cvmfs/cache_stream.cc

```diff
@@ -0,0 +1,320 @@
+  int retval = pthread_mutex_init(lock_fd_table_, NULL);
+  assert(retval == 0);
+
+  delete quota_mgr_;
+  quota_mgr_ = cache_mgr_->quota_mgr();
+}
+
+StreamingCacheManager::~StreamingCacheManager() {
+  pthread_mutex_destroy(lock_fd_table_);
+  free(lock_fd_table_);
+  quota_mgr_ = NULL;  // gets deleted by cache_mgr_
+}
+
+std::string StreamingCacheManager::Describe() {
+  return "Streaming shim, underlying cache manager:\n" + cache_mgr_->Describe();
+}
+
+bool StreamingCacheManager::AcquireQuotaManager(QuotaManager *quota_mgr) {
+  bool result = cache_mgr_->AcquireQuotaManager(quota_mgr);
+  if (result)
+    quota_mgr_ = cache_mgr_->quota_mgr();
+  return result;
+}
+
+int StreamingCacheManager::Open(const LabeledObject &object) {
+  int fd_in_cache_mgr = cache_mgr_->Open(object);
+  if (fd_in_cache_mgr >= 0) {
+    MutexLockGuard lock_guard(lock_fd_table_);
+    return fd_table_.OpenFd(FdInfo(fd_in_cache_mgr));
+  }
+
+  if (fd_in_cache_mgr != -ENOENT)
+    return fd_in_cache_mgr;
+
+  if (object.label.IsCatalog() || object.label.IsPinned() ||
+      object.label.IsCertificate())
+  {
+    return -ENOENT;
+  }
+
+  MutexLockGuard lock_guard(lock_fd_table_);
+  return fd_table_.OpenFd(FdInfo(object));
+}
+
+int64_t StreamingCacheManager::GetSize(int fd) {
+  FdInfo info;
+  {
+    MutexLockGuard lock_guard(lock_fd_table_);
+    info = fd_table_.GetHandle(fd);
+  }
+
+  if (!info.IsValid())
+    return -EBADF;
+
+  if (info.fd_in_cache_mgr >= 0)
+    return cache_mgr_->GetSize(info.fd_in_cache_mgr);
+
+  return Stream(info, NULL, 0, 0);
+}
+
+int StreamingCacheManager::Dup(int fd) {
+  FdInfo info;
+
+  MutexLockGuard lock_guard(lock_fd_table_);
+  info = fd_table_.GetHandle(fd);
+
+  if (!info.IsValid())
+    return -EBADF;
+
+  if (info.fd_in_cache_mgr >= 0) {
+    int dup_fd = cache_mgr_->Dup(info.fd_in_cache_mgr);
+    if (dup_fd < 0)
+      return dup_fd;
+    return fd_table_.OpenFd(FdInfo(dup_fd));
+  }
+
+  return fd_table_.OpenFd(FdInfo(LabeledObject(info.object_id, info.label)));
+}
+
+int StreamingCacheManager::Close(int fd) {
+  FdInfo info;
+  {
+    MutexLockGuard lock_guard(lock_fd_table_);
+    info = fd_table_.GetHandle(fd);
+    if (!info.IsValid())
+      return -EBADF;
+    fd_table_.CloseFd(fd);
+  }
+
+  if (info.fd_in_cache_mgr >= 0)
+    return cache_mgr_->Close(info.fd_in_cache_mgr);
+
+  return 0;
+}
+
+int64_t StreamingCacheManager::Pread(
+  int fd, void *buf, uint64_t size, uint64_t offset)
+{
+  FdInfo info;
+  {
+    MutexLockGuard lock_guard(lock_fd_table_);
+    info = fd_table_.GetHandle(fd);
+  }
+
+  if (!info.IsValid())
+    return -EBADF;
+
+  if (info.fd_in_cache_mgr >= 0)
+    return cache_mgr_->Pread(info.fd_in_cache_mgr, buf, size, offset);
+
+  uint64_t nbytes_streamed = Stream(info, buf, size, offset);
+  if (nbytes_streamed < offset)
+    return 0;
+  if (nbytes_streamed > (offset + size))
+    return static_cast<int64_t>(size);
+  return static_cast<int64_t>(offset + size - nbytes_streamed);
```

Comment:
Not sure I understand this line. Why is nbytes_streamed subtracted here?

Reply:
Thanks, that's a good catch!  We should be returning `nbytes_streamed - offset`. Patch coming.